From f7ca80f3d31643ac7cfbe5a370d5359b9534729c Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Wed, 07 Oct 2020 23:05:47 +0300
Subject: [PATCH] [SW-12754] gaudi: security: add hbw rrs protection

Add HBW protection with Range Registers.
The HBW RRs protection in Gaudi is implemented by using "base and mask"
concept. Therefore some restrictions are have to be coped with.
Specifically, a "base" implies the maximal protection size (the "mask")
per range register and thus we are forced to use several range registers
in some of contiguous areas (e.g. SRAM, SRAM reduction).

Note: The HBM and SRAM protection is disabled until the dynamic HBW
protection is implemented. This prevents bricking devices.

Change-Id: Iad6c956eb7d9f2275d66ce3cb16afe303326c355
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index d32e514..8c3e987 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -22,6 +22,8 @@
 #include <nic0_txs0_regs.h>
 #include <nic0_txe0_regs.h>
 #include <nic0_tmr_regs.h>
+#include <sif_rtr_0_regs.h>
+#include <sif_rtr_ctrl_0_regs.h>
 #include <sif_rtr_ctrl_7_regs.h>
 #include <nif_rtr_0_regs.h>
 #include <nif_rtr_ctrl_0_regs.h>
@@ -838,6 +840,85 @@
 	 */
 }
 
+static void preboot_secure_hbw_rr(int rr, u64 base, u64 mask)
+{
+	int i;
+	u32 base_l = (u32)base, base_h = (u32)(base >> BITS_IN_REG);
+	u32 mask_l = (u32)mask, mask_h = (u32)(mask >> BITS_IN_REG);
+	u32 rr_base, rr_dest, hit;
+
+	for (i = 0; i < IF_RTRS_NUM; i++) {
+		rr_base = rr * sizeof(u32) + IF_RTRS_DIST * i;
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AW_0;
+		regmap_writel(base_l, rr_dest);
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_BASE_HIGH_AW_0;
+		regmap_writel(base_h, rr_dest);
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_MASK_LOW_AW_0;
+		regmap_writel(mask_l, rr_dest);
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_MASK_HIGH_AW_0;
+		regmap_writel(mask_h, rr_dest);
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_BASE_LOW_AR_0;
+		regmap_writel(base_l, rr_dest);
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_BASE_HIGH_AR_0;
+		regmap_writel(base_h, rr_dest);
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_MASK_LOW_AR_0;
+		regmap_writel(mask_l, rr_dest);
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_MASK_HIGH_AR_0;
+		regmap_writel(mask_h, rr_dest);
+
+		/* Enable (hit) RR */
+		rr_base = IF_RTRS_DIST * i;
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_HIT_AW;
+		hit = regmap_readl(rr_dest);
+		regmap_writel(hit | BIT(rr), rr_dest);
+
+		rr_dest = rr_base + mmSIF_RTR_CTRL_0_RANGE_PRIV_HIT_AR;
+		hit = regmap_readl(rr_dest);
+		regmap_writel(hit | BIT(rr), rr_dest);
+	}
+}
+
+static void preboot_secure_hbw(void)
+{
+	/* HBM RR_0 0 - 32GB */
+	/*
+	 * TODO: The below is commented out temporarily until the dynamic
+	 * DRAM RR protection is in place.
+	 */
+/*	preboot_secure_hbw_rr(0, 0, 0x3FFF800000000);
+*/
+	/* SRAM RR_8 0x7F_F000_0000 - 0x7F_F100_0000 (16MB) */
+	/*
+	 * TODO: The below (both SRAM parts) are commented out temporarily
+	 * until the dynamic SRAM RR protection is in place.
+	 */
+/*	preboot_secure_hbw_rr(8, 0x7FF0000000, 0x3FFFFF000000);
+*/
+	/* SRAM RR_9 0x7F_F100_0000 - 0x7F_F140_0000 (4MB) */
+/*	preboot_secure_hbw_rr(9, 0x7FF1000000, 0x3FFFFFC00000);
+*/
+	/* SRAM Reduct. RR_1 0x7F_F140_0000 - 0x7F_F180_0000 (4MB) */
+	preboot_secure_hbw_rr(1, 0x7FF1400000, 0x3FFFFFC00000);
+
+	/* SRAM Reduct. RR_2 0x7F_F180_0000 - 0x7F_F200_0000 (8MB) */
+	preboot_secure_hbw_rr(2, 0x7FF1800000, 0x3FFFFF800000);
+
+	/* SRAM Reduct. RR_3 0x7F_F200_0000 - 0x7F_F280_0000 (8MB) */
+	preboot_secure_hbw_rr(3, 0x7FF2000000, 0x3FFFFF800000);
+
+	/* STM RR_4 0x7F_F400_0000 - 0x7F_F800_0000 (64MB) */
+	preboot_secure_hbw_rr(4, 0x7FF4000000, 0x3FFFFC000000);
+
+	/* SPI Flash RR_5 0x7F_F800_0000 - 0x7F_FA00_0000 (32MB) */
+	preboot_secure_hbw_rr(5, 0x7FF8000000, 0x3FFFFE000000);
+
+	/* Scratchpad SRAM RR_6 0x7F_FBFE_0000 - 0x7F_FBFF_0000 (64KB) */
+	preboot_secure_hbw_rr(6, 0x7FFBFE0000, 0x3FFFFFFF0000);
+
+	/* PCIe FW SRAM RR_7 0x7F_FBFF_0000 - 0x7F_FBFF_8000 (32KB) */
+	preboot_secure_hbw_rr(7, 0x7FFBFF0000, 0x3FFFFFFF8000);
+}
+
 static void preboot_security_pci_disable(void)
 {
 	u32 hit;
@@ -903,6 +984,8 @@
 	preboot_secure_if();
 	preboot_secure_sram_cfg();
 
+	preboot_secure_hbw();
+
 	/* TODO: implement the rest of security enablement */
 
 	cpu_boot_err_clrset(CPU_BOOT_ERR0_SECURITY_NOT_RDY, false);
